### PR TITLE
Fill gaps in System.Data unit testing

### PR DESCRIPTION
**Pull request by logiclrd**
This PR backfills missing automated tests for System.Data.

## IMPORTANT 

* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by unit tests which follow the Arrange-Act-Assert syntax and the naming conventions such as is used [in these tests](../tree/develop/Tests/FluentAssertions.Equivalency.Specs/MemberMatchingSpecs.cs#L51-L430).
* [x] If the PR adds a feature or fixes a bug, please update [the release notes](../tree/develop/docs/_pages/releases.md) with a functional description that explains what the change means to consumers of this library, which are published on the [website](https://fluentassertions.com/releases).
* [x] If the PR changes the public API the changes needs to be included by running [AcceptApiChanges.ps1](../tree/develop/AcceptApiChanges.ps1) or [AcceptApiChanges.sh](../tree/develop/AcceptApiChanges.sh).
* [x] If the PR affects [the documentation](../tree/develop/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).

### CODE REVIEW

File: Tests/FluentAssertions.Equivalency.Specs/DataRowSpecs.cs [hunk 1]
```diff
@@ -1,332 +1,421 @@
...
 
-                var dataSet2 = new TypedDataSetSubclass(dataSet1);
+            // Act & Assert
+            dataRow1.Should().BeEquivalentTo(dataRow2, config => config.ExcludingColumn(dataSet2.TypedDataTable1.DateTimeColumn));
+        }
 
-                var dataRow1 = dataSet1.TypedDataTable1[0];
-                var dataRow2 = dataSet2.TypedDataTable1[0];
+        [Fact]
+        public void When_data_has_column_asserting_HaveColumn_should_succeed()
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Equivalency.Specs/DataRowSpecs.cs:336]: ❌ Keep the names functional and concise without using the literal names of code elements. Same for all the other tests.

File: Tests/FluentAssertions.Equivalency.Specs/DataRowSpecs.cs [hunk 2]
```diff
@@ -1,332 +1,421 @@
...
+            // Arrange
+            var dataSet = CreateDummyDataSet<TypedDataSetSubclass>();
 
-                dataRow1.DateTime = DateTime.UtcNow;
-                dataRow2.DateTime = dataRow1.DateTime;
+            var dataRow = dataSet.TypedDataTable1[0];
 
-                dataSet2.AcceptChanges();
+            // Act & Assert
+            dataRow.Should().HaveColumn("ForeignRowID");
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Equivalency.Specs/DataRowSpecs.cs:344]: 🤔 How do I know from the test that it doesn't have that column?

File: Tests/FluentAssertions.Equivalency.Specs/DataRowSpecs.cs [hunk 3]
```diff
@@ -1,332 +1,421 @@
...
-            {
-                // Arrange
-                var dataSet1 = CreateDummyDataSet<TypedDataSetSubclass>();
+            // Act
+            Action action =
+                () => dataRow.Should().HaveColumn("Unicorn");
 
-                var dataSet2 = new TypedDataSetSubclass(dataSet1);
+            // Assert
+            action.Should().Throw<XunitException>();
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Equivalency.Specs/DataRowSpecs.cs:360]: 🤔 How do I know the exception was raised for the right reasons?
🤔 How do I know if that column exists or not from this test?

File: Tests/FluentAssertions.Equivalency.Specs/DataRowSpecs.cs [hunk 4]
```diff
@@ -1,332 +1,421 @@
...
-                dataSet2.AcceptChanges();
+            var columnNames = dataRow.Table.Columns.OfType<DataColumn>()
+                .Take(3)
+                .Select(column => column.ColumnName);
 
-                // Act & Assert
-                dataRow1.Should().BeEquivalentTo(dataRow2, config => config.Excluding(dataRow => dataRow.RowState));
-            }
+            // Act & Assert
+            dataRow.Should().HaveColumns(columnNames);
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Equivalency.Specs/DataRowSpecs.cs:376]: 🤔 I don't get this assertion. Why are you _taking_ only 3 and then expect this to match to the total number of rows?

File: Tests/FluentAssertions.Equivalency.Specs/DataSetSpecs.cs [hunk 5]
```diff
@@ -610,5 +612,108 @@ public void When_Tables_contain_different_data_it_should_fail()
             // Assert
             action.Should().Throw<XunitException>();
         }
+
+        [Fact]
+        public void When_asserting_table_count_if_it_matches_it_should_succeed()
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Equivalency.Specs/DataSetSpecs.cs:617]: 🙃 That's a weird sentence.

**Review by dennisdoomen — CHANGES_REQUESTED** [review]
Thanks for taking the time for this. I stopped repeating myself, so please consider my comments to be applicable on every new test in the PR.

File: Tests/FluentAssertions.Equivalency.Specs/DataSetSpecs.cs (hunk 5 above)
Comment (dennisdoomen) [reply Tests/FluentAssertions.Equivalency.Specs/DataSetSpecs.cs:617]: If C# names supported commas, then no.

File: Tests/FluentAssertions.Equivalency.Specs/DataRowSpecs.cs (hunk 4 above)
Comment (dennisdoomen) [reply Tests/FluentAssertions.Equivalency.Specs/DataRowSpecs.cs:376]: But why 3 specifically? I can't see that from the test. If it's just a subset, I would name `columnNames` something like `subsetOfColumnNames` and take the number of columns in that table minus 1.

**Comment by dennisdoomen** [comment]
@eNeRGy164 why does a PR like this not report code coverage?

**Review by dennisdoomen — APPROVED** [review]

**Comment by eNeRGy164** [comment]
> @eNeRGy164 why does a PR like this not report code coverage?

Because develop is not (recently) pulled into this branch.

File: Tests/FluentAssertions.Equivalency.Specs/DataRowSpecs.cs [hunk 6]
```diff
@@ -1,332 +1,426 @@
...
+            // Arrange
+            var dataSet = CreateDummyDataSet<TypedDataSetSubclass>();
 
-                var dataRow1 = dataSet1.TypedDataTable1[0];
-                var dataRow2 = dataSet2.TypedDataTable1[0];
+            var dataRow = dataSet.TypedDataTable1[0];
 
-                dataRow1.DateTime = DateTime.UtcNow;
-                dataRow2.DateTime = dataRow1.DateTime;
+            string expectedColumnName = dataSet.TypedDataTable1.Columns.OfType<DataColumn>().Last().ColumnName;
```
Comment (jnyrup) [inline Tests/FluentAssertions.Equivalency.Specs/DataRowSpecs.cs:346]: Could this be `.Cast<DataColumn>()` instead?
That avoids unexpectedly leaving any `nulls` out.

I see that `DataRowEquivalencyStep` and `DataSetEquivalencyStep` uses `OfType` as wel..

**Review by jnyrup — COMMENTED** [review]
Thanks for adding the extra tests.
Please rebase on `develop`, so the PR will trigger the newly added code coverage.

File: Tests/FluentAssertions.Equivalency.Specs/DataRowSpecs.cs (hunk 6 above)
Comment (logiclrd) [reply Tests/FluentAssertions.Equivalency.Specs/DataRowSpecs.cs:346]: I don't think it's possible to put a `null` into any of the System.Data collections. `OfType` and `Cast` should be functionally equivalent in this context. I can change them to `Cast`.

**Comment by logiclrd** [comment]
Rebased. :-)

**Review by jnyrup — APPROVED** [review]

**Pull request merged**